### PR TITLE
Make debugBuilding available in release mode

### LINKED ISSUE #34724: The .of pattern needs a mean to not add a dependency if called outside of build/didChangeDependencies
Consider the following `Foo` widget:

```dart
class Foo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return RaisedButton(
      onPressed: () {
        MyInherited.of(context).doSomething();
      }
    );
  }
}
```

In that code snippet, for most implementations of `MyInherited.of`, pressing the RaisedButton will add a dependency `Foo` and `MyInherited`.

Which means after the button is clicked once, updating `MyInherited` will force `Foo` to rebuild, for no reason.

___

It is technically possible to prevent that from happening, using `ancestorInheritedElementForWidgetOfExactType` instead of `inheritFromWidgetOfExactType`.

But the fact that `inheritFromWidgetOfExactType` add the dependency when used outside of `build`/`didChangeDependencies` is probably more of a bug that the desired behavior.

____

There are multiple solutions:

- do a breaking change on `inheritFromWidgetOfExactType`, to not add a dependency when used outside of `build`/`didChangeDependencies`.

- Add a new method on `BuildContext`/`Element` that wraps `inheritFromWidgetOfExactType` and `ancestorInheritedElementForWidgetOfExactType` into something that switch between them depending on if used inside `build`/`didChangeDependencies` or not.

I can't think of a legitimate use-case where we'd want the current behavior. So although breaking, the first solution would probably a very little impact.

Adding a new method on `BuildContext` for the `.of` pattern, although non-breaking, will likely have a much bigger impact.
Because that would implicitly require nearly all existing `.of` methods to be updated to use the new life-cycle.

cc @Hixie

### PR DESCRIPTION
**Pull request by rrousselGit**
## Description

This makes `BuildOwner.debugBuilding` available in release mode.
It therefore allows `Element` to do custom behavior based on whether the widget tree is building or not.

The main use-case is to know when to call `inheritFromElement` or not inside the classical static `.of` method:

```dart
static T of(BuildContext context) {
    final inheritedElement = context.ancestorInheritedElementForWidgetOfExactType(T);
    if (context.owner.isBuilding) {
      context.inheritFromElement(inheritedElement);
    }
    return inheritedElement;
}
```

This PR does not update the existing  `.of` methods to use that logic (as it is for a 3rd party package).
If these changes are accepted, it may be a good idea to apply that logic to built-in InheritedWidgets, like Theme.of/Scrollable.of.

## Related Issues

closes https://github.com/flutter/flutter/issues/34724

## Tests

I added the following tests:

- packages/fluter/test/widgets/build_owner_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*

^ I've read through [Handling breaking changes], but I don't fully understand the process.
Should I mail flutter-announce@googlegroups.com directly?

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/build_owner_test.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            expect(tester.binding.buildOwner.isBuilding, isTrue);
+          },
+          builder: (BuildContext _) {
+            expect(tester.binding.buildOwner.isBuilding, isTrue);
+            return Container();
+          },
+        ),
+      );
+
+      expect(tester.binding.buildOwner.isBuilding, isFalse);
```
Comment (mono0926) [inline packages/flutter/test/widgets/build_owner_test.dart]: How about `initState`?
On that timing, tester.binding.buildOwner.isBuilding should be `isTrue` 🤔

Comment (rrousselGit) [reply packages/flutter/test/widgets/build_owner_test.dart]: That's expected IMO. Many built in Inheritedwidget are currently inaccessible from initState, where it is recommended to use didChangeDependencies instead

**Review by goderbauer — COMMENTED** [review]
I am not convinced that exposing this is such a great idea.

The example use case given scares me a little because a function that changes its behavior based on where it is being called from will be hard to reason about.

**Comment by rrousselGit** [comment]
> The example use case given scares me a little because a function that changes its behavior based on where it is being called from will be hard to reason about.

The current situation is indefinitely worse in my opinion. 
There's _never_ a situation where we want a click handler to add a dependency on an Inheritedwidget. 

I've seen many peoples complain that, in the current situation, it is too easy to mistakenly add an undesired dependency. 

Peoples get confused by that. They don't understand why obtaining a value inside a click handler make the widget rebuild more often.

**Comment by goderbauer** [comment]
Depending on global state and changing a function's behavior based on that still doesn't seem like a good idea.

You could introduce a flag to your static `of` methods that allows the caller to explicitly switch between adding the dependency or not. Or, you could have a separate static method for not registering the dependency.

Orthogonal to this, I am curious to learn why people care about the rebuild-frequency so much that adding this extra dependency becomes an issue. Widget-builds should be fast and side-effect free.

**Comment by rrousselGit** [comment]
> You could introduce a flag to your static of methods that allows the caller to explicitly switch between adding the dependency or not. Or, you could have a separate static method for not registering the dependency.

That's already what `provider` do.
This works fine if peoples use it correctly. But that's the issue: peoples don't.

Multiple users of `provider` have complained that such flag is too easy to forget. And this doesn't count all the users that aren't even aware of such flag.

> Orthogonal to this, I am curious to learn why people care about the rebuild-frequency so much that adding this extra dependency becomes an issue. Widget-builds should be fast and side-effect free.

I can't speak for others, as I, too, think such optimization is not "critical".

But I don't think we can deny that it's still good to take. If it truly didn't matter,  [AnimatedBuilder](https://api.flutter.dev/flutter/widgets/AnimatedBuilder-class.html) wouldn't take an extra `child` argument used exclusively to optimize rebuilds.

And while rebuilding widgets is cheap in itself, the layout/painting phases are more expensive. But an extra undeeded rebuild can trigger mark a branch as needing layout/paint.

**Comment by rrousselGit** [comment]
In any case, if this change is unreasonable then `provider` will likely opt for one of these alternatives:

- Try to fight against the framework to it myself externally. I haven't tested it yet, but it sounds feasible using `addPersistentFrameCallback` and `addPostFrameCallback`.
  But that's hacky, I don't like it.

- make the `.of` method throw in development if the `listen` flag (similar what's described [here](https://github.com/flutter/flutter/pull/40491#issuecomment-532318052)) is not specified correctly.
  Basically we'd have:
  ```dart
  static T of(BuildContext context, {bool listen = true}) {
    assert(listen == context.owner.debugBuilding);
    final inheritedElement = context.ancestorInheritedElementForWidgetOfExactType(T);
    if (listen) {
      context.inheritFromElement(inheritedElement);
    }
    return inheritedElement;
  }
  ```
  This will prevent peoples from making mistakes. But it makes the flag pointless. Why should the function take a parameter if there's always only one valid value and it's known by the function?

**Comment by rrousselGit** [comment]
Here's a StackOverflow question that got asked today about this specific problem: https://github.com/flutter/flutter/pull/40491

The word "counterintuitive" describes quite well why the current issue

**Comment by goderbauer** [comment]
Your second suggestion is what I would probably go with. It's explicit and doesn't rely on global state. Maybe chose a better name than `listen` though. Something that makes it clear that setting this parameter will cause the widget to get rebuilt. Additional, this parameter will need good documentation. From the stackoverflow it sounds more like the user wasn't sure what that parameter does - usually pointing to a documentation problem. Last, but not least, you could make it an `@required` parameter to make people actively think about its value (and hopefully consult the docs if they are unsure) anytime they use the `of` method. Just some ideas...

I am going to close this PR for now because I don't think we want this change in the framework.

**Pull request closed without merging**